File: lm/srilm.py
#---------------------------------------------------------------
# PyNLPl - SRILM Language Model
#   by Maarten van Gompel, ILK, Universiteit van Tilburg
#   http://ilk.uvt.nl/~mvgompel
#   proycon AT anaproy DOT nl
#
#   Adapted from code by Sander Canisius
#
#   Licensed under GPLv3
#
#
# This library enables using SRILM as language model
#
#----------------------------------------------------------------

from __future__ import print_function
from __future__ import unicode_literals
from __future__ import division
from __future__ import absolute_import    

import srilmcc
from pynlpl.textprocessors import Windower

class SRILM:
    def __init__(self, filename, n):
        self.model = srilmcc.LanguageModel(filename, n)
        self.n = n

    def scoresentence(self, sentence, unknownwordprob=-12):
        score = 0
        for ngram in Windower(sentence, self.n, "<s>", "</s>"):
            try:
               score += self.logscore(ngram)
            except KeyError:
               score += unknownwordprob
        return 10**score

    def __getitem__(self, ngram):
        return 10**self.logscore(ngram)

    def __contains__(self, key):
        return self.model.exists( key )

    def logscore(self, ngram):
        #Bug work-around
        #if "" in ngram or "_" in ngram or "__" in ngram:
        #    print >> sys.stderr, "WARNING: Invalid word in n-gram! Ignoring", ngram 
        #    return -999.9

        if len(ngram) == self.n:
            if all( (self.model.exists(x) for x in ngram) ):
                #no phrases, basic trigram, compute directly
                return self.model.wordProb(*ngram)
            else:
                raise KeyError
        else:
            raise Exception("Not an " + str(self.n) + "-gram")

File: lm/srilm.cc
#include <string>

#include <boost/python.hpp>
using namespace boost::python;

#include "srilm/include/File.h"
#include "srilm/include/Ngram.h"
#include "srilm/include/Vocab.h"
//#include <NgramStats.h>
#include "srilm/lm/src/NgramStatsLong.cc"

class LanguageModel
{
private:
	Vocab vocab;
	Ngram model;

public:
	LanguageModel(const std::string& filename, int order) : model(vocab, order)
	{
		File file(filename.c_str(), "r");
		model.read(file);
	}

	LogP wordProb(const std::string& context1, const std::string& context2, const std::string& word)
	{
		const VocabIndex context[] = {
			context2 == "__" ? Vocab_None : vocab.getIndex(context2.c_str()),
			context1 == "__" ? Vocab_None : vocab.getIndex(context1.c_str())
		};

		return model.wordProb(vocab.getIndex(word.c_str()), context);
	}
};


BOOST_PYTHON_MODULE(srilmcc)
{
	class_<LanguageModel, boost::noncopyable>("LanguageModel", init<const std::string&, int>())
		.def("wordProb", &LanguageModel::wordProb)
	;
}
